tests: adding test cases for interrupt

enjiamai opened the pull request:
Add test cases of interrupt for improving test coverage. Some of the test cases for other platform were already existing, so this submission mainly focus on adding the missing portion.

The new test case "test_isr_static" add following tests for all platform except qemu_xtensa:
1. To check parameters pass to interrupt handler.
2. To declare interrupts statically at build time.
3. To validate multiple ISRs utilizing the same function to process interrupts.
4. To validate kernel provide a dedicated interrupt stack for processing hardware interrupts.

And try to add support dynamic interrupt testing for the x86, x86_64, and native_posix platform in original test_dynamic_isr.  

Signed-off-by: Enjia Mai <enjiax.mai@intel.com>

Commit:
tests: improve test cases for interrupt testing

Add and improve dynamic and static interrupt test cases of interrupt
for platform such as x86, x86_64, native_posix, etc.

Signed-off-by: Enjia Mai <enjiax.mai@intel.com>

ioannisg commented on tests/kernel/interrupt/src/interrupt_common.c:
```diff
@@ -0,0 +1,130 @@
+/**
+ * @brief Test interrupt features
+ *
+ * @details This test case validates four features of interrupt
```
This seems to me cross-ARCH; can this be done for the other architectures, or is it done already?

enjiamai replied:
Most parts of this case can be found in other test cases, such as configuration of interrupts at build time/runtime, or pass argument to interrupt handler. I saw them in tests/kernel/interrupt, tests/kernel/gen_isr_table, etc.
 Should I implement this test case to cross-ARCH? Any suggestion for this?

enjiamai replied:
This file name has been updated. Fixed.

ioannisg reviewed (commented):
Currently this seems to be x86-specific, though I don't fully-understand why,
so i leave this to the x86 maintainers

stephanosio reviewed (changes requested):
Overall, needs better abstraction and cross architecture support.

If not, `interrupt_common.c` should really be renamed to something like `interrupt_x86.c` or, even better, be just moved to the arch tests (`tests/arch/x86`).

enjiamai commented:
@stephanosio Ok, got your point, will update all of this ASAP, thanks for your advice!

enjiamai commented on tests/kernel/interrupt/src/interrupt_common.c:
```diff
@@ -0,0 +1,130 @@
+	default:
+		break;
+	}
+}
```
OK, already move this to interrupt_util.h, and implement the trigger_irq for x86.

enjiamai commented on tests/kernel/interrupt/src/interrupt_common.c:
```diff
@@ -0,0 +1,130 @@
+static void get_stack_addr(volatile ulong_t *reg)
+{
+	__asm__ volatile("mov %%esp,%0" : "=m"(*reg) : );
+}
```
Ok, also move this to interrupt_util.h. But not implement for other architecture yet. It takes some time to do this, will update it later.

enjiamai replied:
Yes, fixed it.

enjiamai commented on tests/kernel/interrupt/src/interrupt_common.c:
```diff
@@ -0,0 +1,130 @@
+void test_interrupt_features(void)
+{
+	int vector_num;
+	ulong_t thread_esp;
```
Fix it. Thanks for your reminder!

enjiamai commented on tests/kernel/interrupt/src/interrupt_common.c:
```diff
@@ -0,0 +1,130 @@
+ *   interrupts, which are passed to the interrupt service routine when the
+ *   interrupt happens.
+ * - The kernel support configuration of interrupts statically at build time.
+ * - The kernel support configuration of interrupts dynamically at runtime.
```
@stephanosio I added this to dynamic_isr.c, for testing X86. And do you think other architecture such as posix, risc-v, nios2, arc,should also be covered too?

enjiamai replied:
yes, remove it. It's already in dynamc_isr.c.

enjiamai commented on tests/kernel/interrupt/src/interrupt_common.c:
```diff
@@ -0,0 +1,130 @@
+ * - The user shall be able to supply a word-sized parameter when configuring
+ *   interrupts, which are passed to the interrupt service routine when the
+ *   interrupt happens.
+ * - The kernel support configuration of interrupts statically at build time.
```
Yes, I totally agree with you. I change the case name to static_isr.c and moved the dynamic test out of here. It only tests x86 now. But I will try to expend this case to fit other architecture later. 
Actually,  tests/kernel/gen_isr_table has already implemented some different platform for interrupt testing, but not for all. I am a little bit confusing here. Should I just try to add more architecture support by improving tests/kernel/gen_isr_table as possible?

enjiamai replied:
fixed it. Put some missing case in dynamic_isr.c already.

enjiamai commented on tests/kernel/interrupt/src/interrupt_common.c:
```diff
@@ -0,0 +1,130 @@
+	vector_num = irq_connect_dynamic(IRQ1_LINE, 1, test_handler,
+			(void *)ISR1_ARG, 0);
+
+	zassert_not_equal(vector_num, -1, "wrong vector");
```
Fixed it in another way. Just try to check it's allocation is failed or not.

enjiamai commented:
@stephanosio @ioannisg  I have updated it, could you please help to review it, thanks!

stephanosio commented on tests/kernel/interrupt/src/static_isr.c:
```diff
@@ -0,0 +1,100 @@
+#include <kernel_structs.h>
+#include "interrupt_util.h"
+
+#if defined(CONFIG_X86)
```
The main problem with static ISR test is that we cannot "dynamically" (as in, at run-time) find an available IRQ, and the available IRQ numbers may vary across the SoCs for some archs, so we cannot hard-code them in the test (unless we `#ifdef` every available SoC or introduce some kind of automation scripts to handle this).

In that sense, I understand that this test needs to be restricted to certain supported arch/SoCs, but we should not be hard-coding x86 here in the static ISR test; instead, such constructs should be abstracted/defined in the `interrupt_util.h` (e.g. checking whether `TEST_IRQ_0_LINE` and `TEST_IRQ_1_LINE` are defined).

enjiamai replied:
I make some change, to put the trigger_irq() function in interrupt_util.h. If there are still things need to be improve, pls kindly let me know, thanks.

enjiamai commented on tests/kernel/interrupt/src/dynamic_isr.c:
```diff
@@ -50,6 +54,28 @@ void test_isr_dynamic(void)
 		     _sw_isr_table[i].arg == argval,
 		     "dynamic isr did not install successfully");
 }
+#elif defined(CONFIG_DYNAMIC_INTERRUPTS) && defined(CONFIG_X86)
+void test_isr_dynamic(void)
+{
+	int vector_num;
+
+	/**TESTPOINT: configuration of interrupts dynamically at runtime */
+	vector_num = irq_connect_dynamic(TEST_IRQ_DYN_LINE, 1, dyn_isr,
+					NULL, 0);
+
+	TC_PRINT("vector(%d)\n", vector_num);
+	zassert_true(vector_num > 0,
+			"irq connect dynamic failed");
+
+	zassert_equal(handler_has_run, 0,
+			"should not get parameter before interrupt");
+
+	trigger_irq(TEST_IRQ_DYN_LINE);
+
+	zassert_equal(handler_has_run, 1,
+			"parameter in interrupt is not correct");
+}
+
```
OK, got it. It indeed a little trouble here when I implemented this. Let me think how to make this better.

enjiamai commented on tests/kernel/interrupt/src/static_isr.c:
```diff
@@ -0,0 +1,100 @@
+#define ISR1_ARG	0xb01dface
+
+static ulong_t handler_test_result;
+uintptr_t int_0_stack_pointer;
+uintptr_t int_1_stack_pointer;
```
Yes, it shall be. Fix it.

enjiamai commented on tests/kernel/interrupt/src/static_isr.c:
```diff
@@ -0,0 +1,100 @@
+/* self-defined handler for testing */
+static void test_handler(void *param)
+{
+	handler_test_result = (ulong_t)param;
```
Just understand. Got it fixed.

ioannisg commented:
what is the status of this PR @enjiamai @stephanosio ?

enjiamai commented:
@ioannisg So sorry for missing your commented for almost two weeks. Due to some reason, I am just working back on it these days. 
I considered the suggestion what @stephanosio give, it shouldn't be hard code for static interrupt and make the test cases more common here (arch-independent) is important. But I didn't come up with a better way to do it. So I will just simply use #ifdef to do it.

enjiamai commented on tests/kernel/interrupt/src/dynamic_isr.c:
```diff
@@ -5,16 +5,20 @@
+static unsigned int handler_has_run;
 
-static void dyn_isr(void *arg)
+__unused static void dyn_isr(void *arg)
```
Yes, this "__unused" should not be used here, fixed it.

enjiamai commented on tests/kernel/interrupt/src/dynamic_isr.c:
```diff
@@ -5,16 +5,20 @@
 {
 	ARG_UNUSED(arg);
+	handler_has_run = 1;
 }
```
fixed it.

enjiamai commented:
@stephanosio @ioannisg Sorry for reply late. I have solved all the comments that you offered.
 The most important part is:
1. Adding missing platform that support dynamic ISR. 
2. Static ISR for all platform except xtensa and nios2. For xtensa, I don't know why I can't trigger it successfully, I will keep look at it and submit another PR to fix it. 

I try my best to make the test case arch-neutral as possible, but it seem not be good enough.
Could you please help to review it again? It's very close to code freezing date, I appreciate your comments for improving it~

enjiamai commented:
@dcpleung excuse me, could you please help to review it when you have time, thanks!

dcpleung commented on tests/kernel/interrupt/src/dynamic_isr.c:
```diff
@@ -50,10 +60,74 @@ void test_isr_dynamic(void)
+			"irq connect dynamic failed");
+
+	zassert_equal(handler_has_run, 0,
+			"(%d)should not get parameter before interrupt",
```
This is about interrupt being triggered but should not be. So maybe something like `handler has run before interrupt trigger` which is more descriptive.

enjiamai replied:
Yes, the words I used here is not adequate, just got it fix, thanks~

dcpleung commented on tests/kernel/interrupt/src/dynamic_isr.c:
```diff
@@ -50,10 +60,74 @@ void test_isr_dynamic(void)
+	trigger_irq(irq_dyn_line);
+
+	zassert_equal(handler_has_run, 1,
+			"(%d)parameter in interrupt is not correct",
```
Something like `interrupt triggered but handler has not run`.

enjiamai replied:
Yes, got it fixed, thanks!

dcpleung commented on tests/kernel/interrupt/src/dynamic_isr.c:
```diff
@@ -50,10 +60,74 @@ void test_isr_dynamic(void)
+
+	/**TESTPOINT: pass word-sized parameter to interrupt */
+	zassert_equal(handler_test_result, ISR_DYN_ARG,
+			"parameter in interrupt is not correct");
```
`parameter in handler is not correct`.

Might want to print out the argument too.

enjiamai replied:
Yes, got it fixed, thanks!

dcpleung commented on tests/kernel/interrupt/src/interrupt_util.h:
```diff
@@ -99,6 +99,66 @@ static inline void trigger_irq(int irq)
 }
+
+/*
+ * Trigger an interrupt of x86 need a immediate nubmer, so hard code the
```
s/nubmer/number/

enjiamai replied:
Got it fixed, thanks!

enjiamai commented:
@dcpleung thank for your comments, I has already fixed all of them.

enjiamai commented:
@LeiW000 Could you please help to review again, thanks!

enjiamai commented:
@stephanosio excuse me, could you please help to review it again when you have time? thanks!

andyross commented on tests/kernel/interrupt/src/interrupt_util.h:
```diff
@@ -99,6 +99,66 @@ static inline void trigger_irq(int irq)
+	printk("Triggering irq : %d\n", irq);
+
+	/* trigger self-defined interrupt */
+	switch (irq) {
```
This switch seems silly as the vectors are a simple function of the interrupt numbers.  Note that the other architectures you added don't try to interpret the numbers.   Why not:

```c
static inline void trigger_irq(int irq)
{
    int vector = irq + 32;
    __asm__ volatile("int %0" :: "i"(vector));
}
```

enjiamai replied:
Wow, it works. This stuck me for a long time, you just solve it in few seconds. thanks so much!

andyross commented on tests/kernel/interrupt/src/static_isr.c:
```diff
@@ -0,0 +1,196 @@
+
+	/**TESTPOINT: dedicated interrupt stack for processing interrupt */
+	zassert_true((int_0_stack_pointer - curr_stack_ptr)
+			> CONFIG_MAIN_STACK_SIZE,
```
This doesn't look correct to me.  This is checking that "an address on the interrupt stack" is more than "main stack size" after "an address on this thread's stack".

What if the linker places the interrupt stack **lower** in memory than this stack?  What if the location of curr_stack_ptr is at the high address of the stack (i.e. it's a shallow call depth) and the int_0_stack_pointer is near the bottom, and the interrupt stack is immediately after?

The test here should be that the int_0_stack_pointer lies outside the main stack.  To do that you need to whitebox the thread struct and pull out the stack size and limit fields to check.

enjiamai replied:
Got it, just let me think about it~

enjiamai replied:
@andyross I make some change of the checking logic, try to check if these two stack are overlap or not. It seems work well except qemu_cortex_a53 and native_posix board. And the stack information of qemu_cortex_a53 are :
 
start 0x400110c, size 0x1000, delta 0x0 (got it from &z_main_thread.stack_info) 

------ 0x400110c0 ----- main thread stack start -----
------ 0x4001107c ------ &dummy in interrupt context
------ 0x40010f4c ------ &dummy in thread context

so it's weird, seems like un-reasonable for me,  sorry, I was stuck here and still working on it.
I appreciate your comment if any, you always help me a lot~

andyross commented on tests/kernel/interrupt/src/static_isr.c:
```diff
@@ -0,0 +1,196 @@
+#endif
+
+/* Helper function for getting current stack pointer */
+static inline void get_stack_addr(uintptr_t *reg)
```
Too complicated.  You don't care about precision here, you just want "any address on the local stack".  So do it portably and take the address of a local variable:

```c
static inline void *get_stack_addr(uintptr_t *reg)
{
    int dummy;
    *reg = &dummy;
}
```

enjiamai replied:
You are right, This is a good arch-neutral way to do that, learn a lot again, thanks!

andyross reviewed (changes requested):
Some nitpicks.  And unless I'm misunderstanding the setup there's a logic error in the stack bounds check.

enjiamai commented:
Hi, @andyross I have fixed the logic error in the stack bounds check base on your comment. The way I applied here is to get the start and end address of the stack, then check if those two region overlap or not. Could you please help to review it again when you have time, thanks!

And I also enable the static interrupt test case qemu_xtensa, but only check one interrupt. Because there are only 2 SW interrupt can be used ( I confirm this from zephyr-sdk-0.11.4/xtensa/sample_controller/xtensa-zephyr-elf/include/arch/include/xtensa/config/core-isa.h ), and one has already been used by irq_offload.    

@andrewboie @andyross For the check failed in qemu_cortex_a53, native_posix and nrf52_bsim(posix), due to their stack pointer in interrupt context was not located in interrupt stack, I guess it might be a problem of qemu. I have to say sorry, as far I still can't figure out why and how to fix it at this moment, so I have to skip them now.

I am looking forward to getting any comment from both of you, thanks!

LeiW000 commented:
@enjiamai , what is status of the PR? Waitting for @andyross and @stephanosio to review?

enjiamai commented:
@LeiW000 Yes, I think I have addressed all the comments they gave. But I guess they did not review it again due to I did not split them to a small commit. I will split them up within the week, Thanks!

LeiW000 commented:
@andyross, @stephanosio, could you please take look at the PR again? Enjia have addressed all your comments.

enjiamai commented:
rework in another PR and close it.
